File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        visited = set()
        result = []
        row = col = 0

        directions = [(0,1), (1,0), (0,-1), (-1,0)]
        dir = 0

        def valid_pos(row, col):
            return (0 <= row < m and 0 <= col < n and (row, col) not in visited)
        
        while len(visited) < m * n:
            if valid_pos(row, col):
                visited.add((row,col))
                result.append(matrix[row][col])

                # 현재 방향으로 더 해줌.
                next_row = row + directions[dir][0]
                next_col = col + directions[dir][1]
                
                # 다음 스텝 미래를 장막을 들추고 본다.
                if valid_pos(next_row, next_col):
                    row = next_row
                    col = next_col
                else:
                    dir = (dir+1) % 4
                    row += directions[dir][0]
                    col += directions[dir][1]

        return result

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        set<pair<int, int>> visited;
        vector<int> result;
        int row = 0, col = 0;
        
        vector<pair<int, int>> directions = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        int dir = 0;
        
        auto valid_pos = [&](int row, int col) -> bool {
            return (0 <= row && row < m && 0 <= col && col < n && 
                    visited.find({row, col}) == visited.end());
        };
        
        while (visited.size() < m * n) {
            if (valid_pos(row, col)) {
                visited.insert({row, col});
                result.push_back(matrix[row][col]);
                
                int next_row = row + directions[dir].first;
                int next_col = col + directions[dir].second;
                
                if (valid_pos(next_row, next_col)) {
                    row = next_row;
                    col = next_col;
                } else {
                    dir = (dir + 1) % 4;
                    row += directions[dir].first;
                    col += directions[dir].second;
                }
            }
        }
        
        return result;
    }
};